Update request

gavin-norman-sociomantic opened the pull request.

gavin-norman-sociomantic commented:
Based on #94.

gavin-norman-sociomantic commented:
Initial sketch of client-side handling, for early browsing.

gavin-norman-sociomantic commented:
Added rough node-side implementation.

gavin-norman-sociomantic commented:
TODO: Try out https://github.com/sociomantic-tsunami/dhtproto/issues/16#issuecomment-364053337.

gavin-norman-sociomantic commented:
Updated: compiles and fully documented. Next step: testing.

gavin-norman-sociomantic commented:
>Updated: compiles and fully documented.

Well... compiles in D1 ;) That's good enough for now.

gavin-norman-sociomantic commented:
Testing under way, using https://github.com/sociomantic-tsunami/dhtnode/pull/90.

gavin-norman-sociomantic commented:
Local tests confirm the following:
* The basic Update procedure works.
* Update detects and rejects conflicts (where the record being updated changes while the request is in progress).

Commit:
Add Update common protocol

Commit:
Add node Update protocol

Commit:
Add fake DHT Update implementation

Commit:
Add Update usage examples

gavin-norman-sociomantic commented:
Updated, unblocked, ready for review.

joseph-wakeling-sociomantic commented on src/dhtproto/client/UsageExamples.d:
```diff
@@ -878,6 +878,119 @@ unittest
+                    // copy the original value, without change.
+                    auto received_record = info.received.value;
+                    (*info.received.updated_value).copy(info.received.value);
+                    goto case;
```
Can you explain what this `goto` will do, and why?

gavin-norman-sociomantic replied:
Fall-through to the next case (required by D2).

gavin-norman-sociomantic replied:
Why? So that `log.trace` is called for this notification.

joseph-wakeling-sociomantic replied:
Ah, gotcha.  I don't think I'd ever come across that syntax before, good to know.

gavin-norman-sociomantic replied:
Yeah, it's kinda weird. Implicit fall-through is disallowed in D2.

Commit:
Add tests for Update

gavin-norman-sociomantic commented:
Added a couple of simple tests.

gavin-norman-sociomantic commented:
...and a release note.

gavin-norman-sociomantic commented:
test this please dmd 2.071

david-eckardt-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,329 @@
+
+    public void preSupportedCodeSent ( Const!(void)[] init_payload )
+    {
+        auto ed = this.connection.event_dispatcher();
```
Not needed.

nemanja-boric-sociomantic replied:
^ Not needed.

david-eckardt-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,329 @@
+        auto message =
+            *ed.message_parser.getValue!(MessageType)(this.init_payload);
+        auto channel = ed.message_parser.getArray!(char)(this.init_payload);
+        auto key = *ed.message_parser.getValue!(hash_t)(this.init_payload);
```
Suggest using `ed.message_parser.parseBody`.

david-eckardt-sociomantic replied:
Never mind, you are using the tail of `this.init_payload` below.

david-eckardt-sociomantic replied:
The advantage of `parseBody` is to detect more message protocol errors by ensuring the message has only the expected fields and not additional data. If I understand the protocol for this request correctly, you could use it like this:

```d
auto message = *ed.message_parser.getValue!(MessageType)(this.init_payload);
// ...
with ( MessageType ) switch ( message )
{
    // Normal sequence: read, wait for response, update.
    case GetRecord:
        cstring key;
        ed.message_parser.parseBody(this.init_payload, key);
        response = this.getWaitUpdate(channel, key);
        break;

    // Either an updated record transferred from another node or a
    // new record that did not exist in the DHT before. Just put it.
    case UpdateRecord:
        cstring key;
        hash_t original_hash;
        const!(void)[] new_value;
        // Read original hash and new value here rathern than in `update`
        ed.message_parser.parseBody(this.init_payload, key, original_hash, new_value);
        response = this.update(channel, key, original_hash, new_value);
        break;

    default:
        ed.shutdownWithProtocolError("Unexpected message from client");
}
```

david-eckardt-sociomantic replied:
I forgot the channel in the above example, but you'll get the point.

nemanja-boric-sociomantic replied:
Ping on `parseBody`

gavin-norman-sociomantic replied:
It can't be used here, as the payload is passed on to `update`, where more fields are read.

nemanja-boric-sociomantic replied:
All clear, thanks!

david-eckardt-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,329 @@
+                        payload.addCopy(MessageType.RecordValue);
+                    }
+                );
+                ed.flush();
```
Flushing seems not to be necessary here.

david-eckardt-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,329 @@
+
+                // TODO: The value could be sent in the same message as the
+                // response. For simplicity, we send them separately, at
+                // present.
```
Why is this simpler?

gavin-norman-sociomantic replied:
It saves having to interpret one message in different ways, on the client side, depending on the contents.

david-eckardt-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,329 @@
+                    payload.addCopy(MessageType.NoRecord);
+                }
+            );
+            ed.flush();
```
Why not just `return MessageType.NoRecord`?

david-eckardt-sociomantic replied:
It looks like you don't want to end the request, the code comment is wrong then.

gavin-norman-sociomantic replied:
Not sure. This looks like a bug, because it shouldn't execute the following section, though it appears that it still will.

david-eckardt-sociomantic replied:
It could be left to the client to decide what to do in this case and respond either with a record value to put in the DHT or a message to cancel the request. This would increase the complexity of the request but be user friendly, avoiding a subsequent `Put` request if the record doesn't exist.

gavin-norman-sociomantic replied:
Yeah, I deliberately chose to leave this as a future enhancement. (An earlier version of Update actually did handle this extra logic, but I removed it to simplify things. I guess this is where this bug / inconsistency comes from.)

david-eckardt-sociomantic commented on src/fakedht/neo/request/Update.d:
```diff
@@ -0,0 +1,116 @@
+    {
+        auto value_in_channel = global_storage.getCreate(channel).get(key);
+        if ( value_in_channel !is null )
+            dg(value_in_channel);
```
Can `value_in_channel` be `null`?

david-eckardt-sociomantic commented on relnotes/update.feature.md:
```diff
@@ -0,0 +1,10 @@
+
+`dhtproto.client.DhtClient`, `dhtproto.client.request.Update`
+
+The new request, `Update`, fetches a record value, allows the user to speficy an
```
Typo `specify`

Commit:
Add Update release note

gavin-norman-sociomantic commented:
Updated:
* Fixed bug around `NoRecord`.
* Refactored code to send/receive the message type and record value in a single payload. (Removed TODOs.)
* Fixed typo in release notes.

gavin-norman-sociomantic commented:
Updated:
* Use `parseBody` in node-side `update` method.

nemanja-boric-sociomantic commented on src/dhtproto/client/NotifierTypes.d:
```diff
@@ -56,6 +56,50 @@ public struct RequestRecordInfo
+    /// Template method to deserialize `value` as a given struct.
+    mixin DeserializeMethod!(value);
+
+    // TODO add serialize method
```
Unresolved TODO.

nemanja-boric-sociomantic replied:
Unresolved TODO.

nemanja-boric-sociomantic commented on src/dhtproto/client/mixins/NeoSupport.d:
```diff
@@ -977,6 +1009,61 @@ template NeoSupport ( )
+        ***********************************************************************/
+
+        public void update ( cstring channel, hash_t key,
+            Neo.Update.Notifier user_notifier )
```
Do we want this one to return `bool`?

gavin-norman-sociomantic replied:
That would inconsistent with the other blocking requests in this class.

nemanja-boric-sociomantic commented on src/dhtproto/client/request/Update.d:
```diff
@@ -0,0 +1,129 @@
+                meantime. The Update request is rejected and has no effect. (The
+                user may retry the request.)
+            4b. The record has not been modified, in the meantime, and the new
+                value is written to the DHT.
```
`The record was deleted or has not been modified`, to be pedant.

gavin-norman-sociomantic replied:
Mm, which sentence is this an addition to?

nemanja-boric-sociomantic replied:
I forgot how the code looked like (I need to check), but AFAIR new value gets written in two cases:

1. the old record value is still valid
2. record ceased to exist in the meantime

nemanja-boric-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,312 @@
+                exists = true;
+                stored_hash = Fnv1a(value);
+            }
+        );
```
I hope there's a guarantee there's no fiber switch between this delegate executing and the next line. Because if there's one, you end up with the race condition within this small window.

gavin-norman-sociomantic replied:
Hm, it depends on the storage engine's implementation. In the two extant implementations, no fiber switch can occur.

nemanja-boric-sociomantic replied:
Right. Perhaps is enough just to add a comment here.

nemanja-boric-sociomantic reviewed (commented):
Some initial comments.

gavin-norman-sociomantic commented:
Updated:
* Added serialize method to Update notifier.
* Added usage example and test for serialization methods.
* Mentioned deleted records in comment.

joseph-wakeling-sociomantic commented on src/dhtproto/client/NotifierTypes.d:
```diff
@@ -89,3 +134,65 @@ public struct RequestKeyInfo
+            if ( client_serializer_version_decorator is null )
+                client_serializer_version_decorator = new VersionDecorator;
+
+            client_serializer_version_decorator.store!(T)(src, *dst);
```
I don't think you actually need to bother with allocating a singleton `VersionDecorator` instance: `VersionDecorator.store` will do, as it is a `static` method:
https://github.com/sociomantic-tsunami/ocean/blob/e9b693dc46d6d6b80bd54244b004b0ecd7679c64/src/ocean/util/serialize/contiguous/MultiVersionDecorator.d#L107

The underlying buffer of `VersionDecorator` is only used for deserialization IIRC.

gavin-norman-sociomantic replied:
Ah that's good to know :+1: .

gavin-norman-sociomantic replied:
Updated.

nemanja-boric-sociomantic replied:
This doesn't seem like it's updated?

joseph-wakeling-sociomantic commented on src/dhtproto/client/NotifierTypes.d:
```diff
@@ -89,3 +134,65 @@ public struct RequestKeyInfo
+
+*******************************************************************************/
+
+public VersionDecorator client_serializer_version_decorator;
```
In line with remarks above about `VersionDecorator.store`, this should not be necessary, so you don't even need to bother to deprecate it ;-)

nemanja-boric-sociomantic replied:
ditto - this is not needed if you use `VersionDecorator.store`

joseph-wakeling-sociomantic commented on relnotes/update.feature.md:
```diff
@@ -0,0 +1,10 @@
+updated value, and replaces the old value in the node with the new value. Note
+that the request will notice if the value being updated has been modified by
+another request, while the Update is in progress. If this happens, the Update
+will be rejected (the client should retry the request).
```
Feels like the `{De}SerializeMethod` stuff should be discussed here (not necessarily in detail, but the principle of what is available).

gavin-norman-sociomantic replied:
Seems a bit specific to me. My feeling is that details of usage should be (and are) covered by the usage examples.

joseph-wakeling-sociomantic replied:
TBH I am a little bit mystified as to why swarm/dhtproto are even caring about (de)serialization.  I suppose it makes it simpler to define the delegates without having to provide them a manually-defined context where deserialization functionality is available.  But OTOH it takes us away from the historical assumption that it's for the user to care about what values are, not the swarm protocols.

Is there something specific to the update request that requires this?

gavin-norman-sociomantic replied:
Nope, the protocol doesn't care. It's just a convenience for the user, who would have to implement this anyway. The other DHT neo requests have the same feature.

joseph-wakeling-sociomantic commented on src/dhtproto/client/NotifierTypes.d:
```diff
@@ -89,3 +134,43 @@ public struct RequestKeyInfo
+    public void serialize ( T ) ( T src )
+    {
+        static if ( Version.Info!(T).exists )
+            VersionDecorator.store!(T)(src, *dst);
```
BTW, with respect to the singleton `VersionDecorator` instance used by `swarm.neo.client.mixins.DeserializeMethod`: given that only `loadCopy` is used, it's not actually necessary for the singleton instance to be exposed directly.  It should suffice to have a `public` function that uses a `private VersionDecorator` instance.

gavin-norman-sociomantic replied:
:+1: that would be a slight improvement.

gavin-norman-sociomantic replied:
(A question for swarm, of course. Not related to this PR.)

nemanja-boric-sociomantic commented on src/dhtproto/client/NotifierTypes.d:
```diff
@@ -89,3 +134,65 @@ public struct RequestKeyInfo
     }
 }
+
+import ocean.util.serialize.contiguous.MultiVersionDecorator;
```
Perhaps move this import to the top of the file.

nemanja-boric-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,314 @@
+    Update request protocol.
+
+    Copyright:
+        Copyright (c) 2018 sociomantic labs GmbH. All rights reserved.
```
RIP

gavin-norman-sociomantic replied:
I'll leave this for a batch transformation.

nemanja-boric-sociomantic commented on src/dhtproto/node/neo/request/Update.d:
```diff
@@ -0,0 +1,314 @@
+
+        Reads a single record from the storage engine. Note that the
+        implementing class must guarantee that no fiber switch can occur during
+        this method.
```
:+1:

nemanja-boric-sociomantic commented on src/dhtproto/client/NotifierTypes.d:
```diff
@@ -168,31 +168,9 @@ template SerializeMethod ( alias dst )
-
-*******************************************************************************/
-
-public VersionDecorator client_serializer_version_decorator;
```
Ah, this all looks like it's in the wrong commit, that's why I didn't notice it before.

nemanja-boric-sociomantic commented on src/dhtproto/client/request/internal/Update.d:
```diff
@@ -0,0 +1,792 @@
+                            error = true;
+                            break;
+                        case Retry:
+                            break;
```
Perhaps add a comment here:

```
// No record on this node, retry on the next one
```

nemanja-boric-sociomantic commented on src/dhtproto/client/request/internal/Update.d:
```diff
@@ -0,0 +1,792 @@
+                            // value (Error) or may have been set to a specific
+                            // code.
+                            error = true;
+                            break;
```
It's weird that this `Abort` actually doesn't abort anything - perhaps we should rename it to something else?

nemanja-boric-sociomantic commented on src/dhtproto/client/request/internal/Update.d:
```diff
@@ -0,0 +1,792 @@
+        response code.
+
+        Params:
+            conn = connection to communicate over
```
`connection to communicate with the node that originally had the record`?

nemanja-boric-sociomantic commented on src/dhtproto/client/request/internal/Update.d:
```diff
@@ -0,0 +1,792 @@
+                n.node_error = RequestNodeInfo(this.context.request_id,
+                    conn.remote_address);
+                Update.notify(this.context.user_params, n);
+                return false;
```
Should we maybe add a `FIXME` here - what happens if we write the new record to the newest node, but we fail to remove it from the old node (perhaps because of the connection issue)? Will we get in any kind of trouble for leaving the record on both nodes? Should we remove it from the new node as well?

gavin-norman-sociomantic replied:
It won't be a problem because when the redistribution process moves the record from the old -> new node, it will be discarded as the record already exists on the new node.

nemanja-boric-sociomantic replied:
That's good, thanks for the clarification!

nemanja-boric-sociomantic commented on src/dhtproto/client/UsageExamples.d:
```diff
@@ -878,6 +878,191 @@ unittest
+                    deserialized.daily_totals[0]++;
+
+                    // Serialize the updated record.
+                    info.received.serialize(*deserialized);
```
Nit pick: It feels a bit odd that you do not reference updated_value here.

gavin-norman-sociomantic replied:
You mean `info.received.updated_value`? There's no need to, if you use the `serialize` method.

nemanja-boric-sociomantic replied:
Yes, but it did caught me off guard - it _feels_ like you're serializing into the received record again, and not it's updated value. I'm not sure what's best here - I think some kind of naming where this `received` would be just `record` _might_ be better.

gavin-norman-sociomantic replied:
Hmm, yeah, `received` here represents the notification "we have received a record" (i.e. the action), rather than the thing that has been received. I don't think it's worth changing the status quo for the sake of this one odd case.

nemanja-boric-sociomantic replied:
Hm right, I forgot about that. 👍

nemanja-boric-sociomantic reviewed (commented):
In general LGTM, although I have some questions/suggestions.

Commit:
Add client Update protocol and assignment method

Commit:
Add request notification struct to provide a value and get an updated version

gavin-norman-sociomantic commented:
Updated.

nemanja-boric-sociomantic commented:
I've restarted the build, some issues with the apt repos are in play.
